enarx-prstate

```diff
@@ -25,19 +25,51 @@ def handle(issue, pr, responsible):
     if needed:
         pr.add_to_assignees(*needed)
 
-# If there are any changes requested, the author is responsible.
-for issue in github.search_issues(f"org:enarx is:pr is:public is:open -is:draft review:changes_requested"):
+# Look through all PRs that have a changes requested review present.
+for issue in github.search_issues(f"org:enarx is:pr is:public is:open -is:draft"):
     pr = issue.as_pull_request()
 
-    handle(issue, pr, {pr.user.login})
+    # Get the latest actionable reviews for all PR reviewers.
+    latest_reviews = {r.user.id: r for r in pr.get_reviews() if r.state == "CHANGES_REQUESTED" or r.state == "APPROVED"}
+    # Then, split that set into Approving Reviewers and Changes Requested reviews.
+    approving_reviewers = {r.user.id: r.user.login for r in latest_reviews if r.state == "APPROVED"}
+    changes_requested_reviews = {r.user.id: r for r in latest_reviews if r.state == "CHANGES_REQUESTED"}
 
-# Otherwise, requested reviewers and reviewers with dismissed reviews are.
-# This handles both where the author requests re-review and where, after
-# approvals, reviews were dismissed due to a new push.
-for issue in github.search_issues(f"org:enarx is:pr is:public is:open -is:draft -review:changes_requested"):
-    pr = issue.as_pull_request()
+    # Collect reviewers whose Changes Requested reviews have been addressed in a set.
+    addressed_reviewers = set()
+    for event in pr.get_issue_events():
+        if event.requested_reviewer is not None and event.requested_reviewer.id in changes_requested_reviews and event.created_at > changes_requested_reviews[event.requested_reviewer.id].submitted_at:
+            # If the above condition is true, that means the PR author has
+            # re-requested a review from the reviewer after the reviewer's last
+            # CHANGES_REQUESTED review was submitted. That implies the PR is
+            # ready for re-review from that reviewer.
+            addressed_reviewers.add(event.requested_reviewer.login)
+
+    # Figure out how many unaddressed reviews there are.
+    unaddressed_reviews = len(changes_requested_reviews) - len(addressed_reviewers)
 
+    # Get all reviewers whose reviews were dismissed. They will need to be
+    # assigned regardless.
     dismissed = {r.user.login for r in pr.get_reviews() if r.state == 'DISMISSED'}
-    requested = {r.login for r in pr.get_review_requests()[0]}
 
-    handle(issue, pr, dismissed | requested)
+    # If the latest review submitted by everyone who submitted a review is an
+    # approving review, then it should be assigned to the Enarx bot for merge.
+    if len(latest_reviews) == len(approving_reviewers):
+        enarxbot = github.get_user("enarxbot")
+        handle(issue, pr, enarxbot.login)
```

Comment:
This doesn't handle the case where there are too few reviews (i.e. 1 on a repo that requires 2). We should probably just check if the pr is mergeable here.

Also, isn't the `enarxbot` login also `enarxbot`? I'm not sure why the query here.

Reply:
I'm actually questioning the utility of assigning `enarxbot` at all.

`prstate` runs immediately after `merge`. That means any PR that was ready for merge should already have gotten merged anyway. And, if anything, that action would be the ideal place to assign the PR author to ensure rebases happen.

Thoughts? Not having to think about that would _really_ simplify the assignment logic.